discovery/v1-generated.ts

```diff
@@ -0,0 +1,2602 @@
+   * @param {String} [options.username] - The username used to authenticate with the service. Username and password credentials are only required to run your application locally or outside of Bluemix. When running on Bluemix, the credentials will be automatically loaded from the `VCAP_SERVICES` environment variable.
+   * @param {String} [options.password] - The password used to authenticate with the service. Username and password credentials are only required to run your application locally or outside of Bluemix. When running on Bluemix, the credentials will be automatically loaded from the `VCAP_SERVICES` environment variable.
+   * @param {Boolean} [options.use_unauthenticated] - Set to `true` to avoid including an authorization header. This option may be useful for requests that are proxied.
+   * @param {Object} [options.headers] - Default headers that shall be included with every request to the service.
+   * @param {Object} [options.headers.X-Watson-Learning-Opt-Out] - Set to `true` to opt-out of data collection. By default, all IBM Watson services log requests and their results. Logging is done only to improve the services for future users. The logged data is not shared or made public. If you are concerned with protecting the privacy of users' personal information or otherwise do not want your requests to be logged, you can opt out of logging.
+   * @constructor
+   * @returns {GeneratedDiscoveryV1}
+   * @throws {Error}
+   */
+  constructor(options: GeneratedDiscoveryV1.Options) {
+    super(options);
+    // check if 'version_date' was provided
+    if (typeof this._options.version_date === 'undefined') {
+      throw new Error(
+        'Argument error: version_date was not specified, use GeneratedDiscoveryV1.VERSION_DATE_2017_09_01'
+      );
+    }
+    this._options.qs.version = options.version_date;
+  }
+
+  /**
+   * Add an environment.
+   *
+   * Creates a new environment.  You can create only one environment per service instance. An attempt to create another environment results in an error.
+   *
+   * @param {Object} params - The parameters to send to the service.
+   * @param {string} params.name - Name that identifies the environment.
+   * @param {string} [params.description] - Description of the environment.
+   * @param {number} [params.size] - **Deprecated**: Size of the environment.
+   * @param {Function} [callback] - The callback that handles the response.
+   * @returns {ReadableStream|void}
+   */
+  createEnvironment(
+    params: GeneratedDiscoveryV1.CreateEnvironmentParams,
+    callback?: GeneratedDiscoveryV1.Callback<GeneratedDiscoveryV1.Environment>
+  ): ReadableStream | void {
+    const _params = extend({}, params);
+    const _callback = typeof callback === 'function' ? callback : () => {};
+    const requiredParams = ['name'];
+    const missingParams = getMissingParams(_params, requiredParams);
+    if (missingParams) {
+      return _callback(missingParams);
```

Comment:
I think we can have 
```js
if (missingParams) {
      return (typeof callback === 'function') ? callback(missingParams) : null;
}
```

Instead of creating our own internal callback

Reply:
Glenn was suggesting this change to the template

`
{{#hasRequiredParams}}
    const _params = extend({}, params);
    const _callback = (callback) ? callback : () => {};
    {{/hasRequiredParams}}
    {{^hasRequiredParams}}
    const _params = (typeof params === 'function' && !callback) ? {} : extend({}, params);
    const _callback = (typeof params === 'function' && !callback) ? params : () => {};
    {{/hasRequiredParams}}
`

Reply:
I know you were saying that we should do what @glennrfisher  proposed in the adapters instead of the Generated service, but we have to consider that the adapters are going to be dissolved at some point and are likely to be short-lived. Therefore, by including what @glennrfisher proposed in the generated service, we reduce the amount of work we need to do after the adapters are dissolved.